### DISCUSSION TITLE
Linter specific functionality

### DISCUSSION THREAD

**Discussion by andreypopp** [w0rp/ale#3959] [Ideas]
I'm wondering what's the stance on adding functionality to ALE which is specific to some LSP linter.

Specifically ocamllsp implements a set of custom commands ([link to docs](https://github.com/ocaml/ocaml-lsp/tree/master/ocaml-lsp-server/docs/ocamllsp)) and I'd really love to use them with ALE.

I've hacked some into my config but I believe this will be useful for a wider ALE's userbase. Should I propose a pull request with such functionality?

**Comment by andreypopp** [comment]
Posting implementations of `:ALESwitchImplIntf` and `:ALEDocumentSymbol` here — https://gist.github.com/5475488b03e44767ed66b136474e210b

**Comment by hsanson** [comment]
ALE's stance if for @w0rp to decide but my opinion on this is that all contributions that improve ALE's LSP support are welcome.

The switch command seems to be non-standard and only specific of ocaml language server. In this case I would say this should not be part of ALE itself and instead ALE should provide methods that allow easy implementation of such features so users can include them in their configs.  Provide documentation in wiki with known implementations like this one.

The documentSymbol is standard LSP and welcome addition to ALE with some modifications:

 1. Make it generic so it works for all LSPs that support it.
 2. Properly advertise the feature is supported by the client (see ale#lsp#Register())
 3. Properly check if the server supports the feature (see s:UpdateCapabilities()) and check if is supported when calling it. Show error message if server does not support it.
 4. Make it agnostic of FZF. Some users may prefer other plugins such as Telescope instead. For example make it work using preview window by default, similar to ALEFindReferences, and allow users to define a callback that invokes FZF or Telescope instead.  This may require significant work on ALE's core functionality.
